Docs/mut query (#845)
bevyengine/bevy-website: content/learn/book/getting-started/ecs/_index.md

Marvelous!

**Quick Note**: "hello world!" might show up in a different order than it does above. This is because systems run in parallel by default whenever possible.

> If we want to change the names of some people because they get married, for example, we can do this using a mutable query:

Reviewer: ```suggestion
If we want to change the names of some people (perhaps they got married!), for example, we can do this using a mutable query:
```

```rs
fn update_people(mut query: Query<&mut Name, With<Person>> {
    for mut name in &mut query {
        if name.0 == "Elaine Proctor" {
            name.0 = "Elaine Hume".to_string();
            break; // We break out of the loop here, because we don’t need to change any other names
        }
    }
}
```

> The query parameter must be mutable and the components we want to change have to be a mutable reference for this to work.

Reviewer: ```suggestion
The `query` argument must be mutable and the components we want to change have to be a mutable reference (`&mut`) for this to work.
```

Don’t forget to add the system to the {{rust_type(type="struct", crate="bevy_app", name="Update")}} schedule:

```rs
fn main() {
    App::new()
        .add_systems(Startup, add_people)
        .add_systems(Update, hello_world)
        .add_systems(Update, (update_people, greet_people).chain())
        .run();
}
```

> Note that we have used `chain()` on the two systems. This is because we want them two to run in exactly that order. If they weren’t, the name might change after we greet the people. But we don’t add the `hello_world` system to this chain, because it doesn’t matter when it runs.

Reviewer: ```suggestion
Note that we have used `chain()` on the two systems. This is because we want them two to run in exactly that order: with `update_people` occurring before `greet_people`.
If they weren’t, the name might change after we greet the people.
But we don’t add the `hello_world` system to this chain, because it doesn’t matter when it runs.
```

Marvelous!

**Quick Note**: "hello world!" might show up in a different order than it does above. This is because systems run in parallel by default whenever possible.

> If we want to change the names of some people (perhaps they got married!), for example, we can do this using a mutable query:

Reviewer 2: Could be useful to add a title or [anchor/id](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#linking_to_an_element_on_the_same_page) so readers can be referred to this section in particular ; but I wouldn't block on it.

Author: like this?

Marvelous!

**Quick Note**: "hello world!" might show up in a different order than it does above. This is because systems run in parallel by default whenever possible.

If we want to change the names of some people (perhaps they got married!), for example, we can do this using a mutable query:

```rs
fn update_people(mut query: Query<&mut Name, With<Person>> {
    for mut name in &mut query {
        if name.0 == "Elaine Proctor" {
            name.0 = "Elaine Hume".to_string();
            break; // We break out of the loop here, because we don’t need to change any other names
        }
    }
}
```

> The `query` argument must be mutable and the components we want to change have to be a mutable reference (`&mut`) for this to work.

Reviewer 3: I think this might be more readable. What do you think?
```suggestion
We need to make `query` mutable, and use a mutable reference (`&mut`) to the components we want to change.
```

Don’t forget to add the system to the {{rust_type(type="struct", crate="bevy_app", name="Update")}} schedule:

```rs
fn main() {
    App::new()
        .add_systems(Startup, add_people)
        .add_systems(Update, hello_world)

>         .add_systems(Update, (update_people, greet_people).chain())

Reviewer 3: I think this might be a bit misleading. I would have written it this way
```rust
        .add_systems(
            Update,
            (hello_world, (update_people, greet_people).chain()),
        )
```

        .run();
}
```

> Note that we have used `chain()` on the two systems. This is because we want them two to run in exactly that order: with `update_people` occurring before `greet_people`.

Reviewer: ```suggestion
Note that we have used `.chain()` on the two systems. This is because we want them two to run in exactly the order they're listed in the code: with `update_people` occurring before `greet_people`.
```

If they weren’t, the name might change after we greet the people.

> But we don’t add the `hello_world` system to this chain, because it doesn’t matter when it runs. That way Bebmvy can run `hello_world` while the other systems are running.

Reviewer 3: this is technically wrong, as `hello_world` cannot run concurrently with `update_people`, as it accesses `Name` mutably.

> But we don’t add the `hello_world` system to this chain, because it doesn’t matter when it runs. That way Bebmvy can run `hello_world` while the other systems are running.

Reviewer 3: ```suggestion
But we don’t add the `hello_world` system to this chain, because it doesn’t matter when it runs. This way, bevy can run `hello_world` while the other systems are running.
```

### Your First mutable Query

If we want to change the names of some people (perhaps they got married!), for example, we can do this using a mutable query:

```rs
fn update_people(mut query: Query<&mut Name, With<Person>> {
    for mut name in &mut query {
        if name.0 == "Elaina Proctor" {
            name.0 = "Elaina Hume".to_string();
            break; // We break out of the loop here, because we don’t need to change any other names
        }
    }
}
```

We need to make `query` mutable, and use a mutable reference (`&mut`) to the components we want to change.

Don’t forget to add the system to the {{rust_type(type="struct", crate="bevy_app", name="Update")}} schedule:

```rs
fn main() {
    App::new()
        .add_systems(Startup, add_people)
        .add_systems(Update, (hello_world, (update_people, greet_people).chain()))
        .run();
}
```

Note that we have used `.chain()` on the two systems. This is because we want them two to run in exactly the order they're listed in the code: with `update_people` occurring before `greet_people`.
If they weren’t, the name might change after we greet the people.

> But we don’t add the `hello_world` system to the chain, because it doesn’t matter when it runs. This way, bevy can run `hello_world` in parallel while the other systems are running.

Reviewer: ```suggestion
But we don’t add the `hello_world` system to the chain, because it doesn’t matter when it runs. This way, Bevy can run `hello_world` in parallel while the other systems are running.
```

Note that we have used `.chain()` on the two systems. This is because we want them two to run in exactly the order they're listed in the code: with `update_people` occurring before `greet_people`.
If they weren’t, the name might change after we greet the people.

> But we don’t add the `hello_world` system to this chain, because it doesn’t matter when it runs. This way, bevy can run `hello_world` while the other systems are running.

Reviewer: ```suggestion
But we don’t add the `hello_world` system to this chain, because it doesn’t matter when it runs. This way, Bevy can run `hello_world` while the other systems are running.
```
